_source/_posts/2020-12-21-beginners-guide-to-jwt.adoc

```diff
@@ -0,0 +1,191 @@
+
+When you use a JWT, it's usually a JWS. The 'S' (the signature) is the important part and allows the token to be validated.  For the rest of this post, I will talk about the JWS format and walk through decoding an example JWT.
+
+== How JWTs Are Used
+
+OAuth 2.0 identity providers (IdP) commonly use JWTs for https://www.oauth.com/oauth2-servers/access-tokens/[access tokens]. You may have seen an HTTP request with an authorization header that looks like this:
+
+[source,http]
+----
+Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJuYW1lIjoiSm9lIENvZGVyIn0.5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U
+----
+
+NOTE: JWT access tokens are NOT part of the OAuth 2.0 specification, but almost all IdPs support them.
+
+Using a JWT (actually a JWS) allows the token to be validated locally, without making an HTTP request back to the IdP, thereby increasing your application's performance.  Applications can make use of data inside the token, further reducing expensive HTTP calls and database lookups.
+
+== JWT structure
+
+A JWS (the most common type of JWT) contains three parts separated by a dot (`.`). The first two parts (the "header" and "payload") are Base64-URL encoded JSON, and the third is a cryptographic signature.
+
+Let's look at an example JWT:
+
+[source,txt]
+----
+eyJhbGciOiJIUzI1NiJ9.eyJuYW1lIjoiSm9lIENvZGVyIn0.5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U
+----
+
+Breaking this down into the individual sections we have:
+
+[source,sh]
+----
+eyJhbGciOiJIUzI1NiJ9 # header
+.
+eyJuYW1lIjoiSm9lIENvZGVyIn0 # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: If you have a JWT with more than three sections, it's probably a JWE.
+
+Next, each of the first two sections are base64-url decoded:
+
+[source,json]
+----
+{"alg":"HS256"} # header
+.
+{"name":"Joe Coder"} # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: On macOS  (with "coreutils" installed, you can base64-url decode the strings on the command line with: `echo eyJhbGciOiJIUzI1NiJ9 | gbasenc -d --base64url`
+
+The last section in the JWT, the signature, is also base64-url encoded, but it's just binary data; if you try to decode it, you will end up with non-displayable characters:
+
+[source,txt]
+----
+��i�i�￹�￹KN�f￹�￹֪�O�_￹R��￹\�+￹
+----
+
+You can use a tool like `hexdump` to view the signatures content:
+
+[source,sh]
+----
+$ echo "5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U=" | gbasenc -d --base64url | hexdump
+
+0000000 e5 d9 69 ec 69 b3 88 bd 90 4b 4e ac 66 02 b8 9a
+0000010 d6 aa bf 4f f1 5f 8a 05 52 e4 c3 87 5c c7 2b 85
+----
+
+=== JWT Claims
+
+Once you start using JWTs you start hearing the word "claim" everywhere. A JWT claim is a key/value pair in a JSON object.  In the example above, `"name": "Joe Coder"`, the claim key is 'name' and the value is 'Joe Coder`. The value of a claim can be any JSON object.
```

Comment:
It seems like 'name' should be `name` and "Joe Coder" should be surrounded with backticks too.